File: Month-2/Week-6/day-38/c++/main.cpp
/**
 * DAY-38 C++ version
 * 
 * Just run 'make' to compile and run all the tests
 */
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "../../../../dependencies/c++/doctest.h" //https://github.com/onqtam/doctest

/**
 * Determine the most expensive computer keyboard and USB drive that can be purchased with a give budget.
 * @param keyboards Prices of keyboards
 * @param drives Prices of USB drives
 * @param Cost to buy the items, -1 if it is not possible to buy both
 */ 
int getMoneySpent(std::vector<int> keyboards, std::vector<int> drives, int b, int r = -1) {    
    for (auto &k : keyboards) for (auto &d : drives) r = (k + d) > r && (k + d) <= b ? (k + d) : r;
    return r;
}

/**
 * Tests
 */
TEST_CASE("Tests")
{    
    CHECK(getMoneySpent({40,50,60},{5,8,12}, 60) == 58);
    CHECK(getMoneySpent({40,50,60},{18,19,20}, 57) == -1);
}

File: Month-2/Week-6/day-38/Go/main.go
package main

import "fmt"

func main() {

	keyboards := []int{40, 50, 60}
	drives := []int{5, 8, 12}
	budget := 60
	result := getMoneySpent(keyboards, drives, budget)

	fmt.Println(result)
}

func getMoneySpent(keyboardPrices, drivePrices []int, budget int) int {

	maxSpent := -1

	for _, k := range keyboardPrices {

		currentKeyboardPrice := k

		for _, d := range drivePrices {
			currentDrivePrice := d
			itemsCombinedPrice := currentKeyboardPrice + currentDrivePrice

			if itemsCombinedPrice <= budget && itemsCombinedPrice > maxSpent {
				maxSpent = itemsCombinedPrice
			}
		}

	}

	return maxSpent

}
